tools/run-jepsen-tests.sh

```diff
@@ -41,39 +62,93 @@ CLOJURE_BIN=$(which clojure)
+
+# build directory should be prepared before use
+[[ -d $BUILD_DIR ]] || mkdir -p $BUILD_DIR
+
+# git must initialized before use even to any unknown user
+CI_COMMIT_SHA=$(git rev-parse HEAD)
+
+# setup cleanup routine, which removes testing nodes by Terraform
 function cleanup {
-    echo "cleanup"
+    echo "Cleanup running ..."
     rm -f $NODES_FILENAME $SSH_KEY_FILENAME
     [[ -e $TERRAFORM_STATE ]] && \
 	terraform destroy -state=$TERRAFORM_STATE -auto-approve $TERRAFORM_CONFIG
 }
 
+# set signal handler on script interrupts
 trap "{ cleanup; exit 255; }" SIGINT SIGTERM ERR
 
-echo -e "${TF_VAR_ssh_key//_/\\n}" > $SSH_KEY_FILENAME
-chmod 400 $SSH_KEY_FILENAME
-$(pgrep ssh-agent 2>&1 > /dev/null) || eval "$(ssh-agent)"
+# setup SSH home path with needed files
+[[ -d $HOME/.ssh ]] || mkdir -p $HOME/.ssh
+chmod 700 $HOME/.ssh
+
+# remove extra spaces from TF_VAR_ssh_key key at the end of lines
+SSH_KEY=$(echo -e "$TF_VAR_ssh_key" | sed 's# *$##g')
+
+# Create file with SSH private key, add it to SSH agent and
+# setup Terraform mandatory TF_VAR_ssh_key_path with its name.
+echo -e "${SSH_KEY//_/\\n}" >$SSH_KEY_FILENAME
+chmod 600 $SSH_KEY_FILENAME
+# reinitialize SSH agent
+eval "$(ssh-agent -s)"
 ssh-add $SSH_KEY_FILENAME
+# check that SSH key added
+ssh-add -l
+export TF_VAR_ssh_key_path=$SSH_KEY_FILENAME
+export TF_VAR_ssh_key=$(cat $SSH_KEY_FILENAME)
 
-if [[ -z ${CI_JOB_ID} ]]; then
+# setup Terraform mandatory TF_VAR_id variable to run the nodes
+if [[ -n ${CI_JOB_ID} ]]; then
     TF_VAR_id=$CI_JOB_ID
 else
     RANDOM_ID=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 13; echo '')
     TF_VAR_id=TF-$RANDOM_ID
 fi
-
 export TF_VAR_id
-export TF_VAR_ssh_key_path=$SSH_KEY_FILENAME
+
+# setup Terraform mandatory TF_VAR_instance_count with nodes number
 [[ -z ${INSTANCE_COUNT} ]] && TF_VAR_instance_count=1 || TF_VAR_instance_count=$INSTANCE_COUNT
 export TF_VAR_instance_count
 
-[[ -e $HOME/.ssh ]] || (mkdir $HOME/.ssh && touch $HOME/.ssh/known_hosts)
+###########################
+# main part of the script #
+###########################
 
+# 1. initiate Terraform configuration for node setup
 terraform init $TERRAFORM_CONFIG
+
+# 2. setup node and create its configuration in $TERRAFORM_STATE file
 terraform apply -state=$TERRAFORM_STATE -auto-approve $TERRAFORM_CONFIG
+
+# 3. print the schemas of the providers used in the configuration
 terraform providers $TERRAFORM_CONFIG
+
+# 4. read an output variable from a Terraform state file and print the value
 terraform output -state=$TERRAFORM_STATE instance_names
+
+# 5. get IP of the node from Terraform state file
 terraform output -state=$TERRAFORM_STATE -json instance_ips | jq --raw-output '.[]' > $NODES_FILENAME
-pushd $TESTS_DIR && lein run $LEIN_OPTIONS
-popd
+
+# 6. add nodes to SSH known hosts and check that they are reachable by SSH
+for node in $(cat $NODES_FILENAME) ; do
+    ssh -vvv -o "StrictHostKeyChecking=no" -o "BatchMode=yes" -i $SSH_KEY_FILENAME ubuntu@$node hostname
+done
```

Comment:
We don't need to check it. `terraform` creates virtual machines and then uses `ssh` to login to guest oses and make an environment preparation.